Doc/c-api/init.rst

```diff
@@ -855,6 +855,12 @@ code, or when embedding the Python interpreter:
    created, the current thread must not have acquired it, otherwise deadlock
    ensues.
 
+   .. warning::
+      Calling this function from a thread when the interpreter is finalizing
+      will terminate the thread, even if the thread was not created by Python.
```

Comment:
"if the thread was not created by Python" is it really possible to call this function if the Python doesn't have a "Python thread state"?

Reply:
I'm not sure "terminate" is the right word. AFAIR we don't terminate OS threads, we merely never resume them anymore?

Also, I would not make this a warning, just a note. Nobody should actually care about this behaviour, since abrupt thread termination is also what happens if the process exits.

Reply:
Rather than trying to describe which exact behavior you get, I suggest to just say that the function calls "PyThread_exit_thread()" on such case.

Reply:
>I would not make this a warning, just a note. Nobody should actually care about this behaviour, since abrupt thread termination is also what happens if the process exits.

@pitrou People wrapping C/C++ libraries that provide callback mechanisms and allow for this callbacks to call into Python care a lot. One scenario is that the object that keeps the external threads alive is destroyed on shutdown and this joins the callback threads. If the threads try to call back into Python they will be killed and the application will deadlocks as the main thread is waiting for the callback to release some lock.

This scenario is very common when wrapping thread based C++ networking libraries.

Reply:
> is it really possible to call this function if the Python doesn't have a "Python thread state"

Whit this I mean that the thread can be created by some C library, not by Python using the threading module. These threads need of course to be registered with the interpreter to call Python code, but the point is avoiding to be killed when you try to do so.

Reply:
Ok. I would still make it a note, though, because it's not a latent security issue or anything close to that.

Reply:
> Ok. I would still make it a note, though, because it's not a latent security issue or anything close to that.

Agreed 👍